src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,105 @@
 package core.basesyntax;
 
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+    private static final int RESIZE_FACTOR = 2;
+    private final int capacity;
+    private final int threshold;
+    private Node<K, V>[] table;
+    private int size;
+
+    public MyHashMap() {
+        table = new Node[DEFAULT_CAPACITY];
+        capacity = table.length;
+        threshold = (int) (capacity * LOAD_FACTOR);
+    }
+
+    private static class Node<K, V> {
+        private final int hash;
+        private final K key;
+        private V value;
+        private Node<K, V> next;
+
+        public Node(int hash, K key, V value, Node<K, V> next) {
+            this.hash = hash;
+            this.key = key;
+            this.value = value;
+            this.next = next;
+        }
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (size == threshold) {
+            resize();
+        }
+        int bucketIndex = getBucketIndex(key);
+        Node<K, V> node = new Node<>(bucketIndex, key, value, null);
+        if (table[bucketIndex] != null) {
+            Node<K, V> currentNode = table[bucketIndex];
+            while (currentNode != null) {
+                if (isKeysEquals(bucketIndex, currentNode.hash, key, currentNode.key)) {
+                    currentNode.value = value;
+                    return;
+                }
+                if (currentNode.next == null) {
+                    currentNode.next = node;
+                    size++;
+                    return;
+                }
+                currentNode = currentNode.next;
+            }
+        }
+        table[bucketIndex] = node;
+        size++;
     }
 
     @Override
     public V getValue(K key) {
+        int bucketIndex = getBucketIndex(key);
+        Node<K, V> node = table[bucketIndex];
+        while (node != null) {
+            if (isKeysEquals(bucketIndex, node.hash, key, node.key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private boolean isKeysEquals(int keyHash, int nodeKeyHash, K key, K nodeKey) {
+        return keyHash == nodeKeyHash
+                && (key == nodeKey || (key != null && key.equals(nodeKey)));
+    }
+
+    private int supplementalHash(int hash) {
+        hash ^= (hash >>> 20) ^ (hash >>> 12);
+        return hash ^ (hash >>> 7) ^ (hash >>> 4);
```

Comment:
What do all these operations mean?

Reply:
guarantees that the received hash codes will only have a limited number of collisions